compiler/rustc_data_structures/src/sync/once_vec.rs

```diff
@@ -0,0 +1,229 @@
+use std::alloc::Layout;
+use std::marker::PhantomData;
+use std::mem::MaybeUninit;
+use std::ptr::NonNull;
+use std::sync::Mutex;
+use std::sync::atomic::{AtomicPtr, AtomicU8, Ordering};
+
+/// Provides a singly-settable Vec.
+///
+/// This provides amortized, concurrent O(1) access to &T, expecting a densely numbered key space
+/// (all value slots are allocated up to the highest key inserted).
+pub struct OnceVec<T> {
+    // Provide storage for up to 2^35 elements, which we expect to be enough in practice -- but can
```

Comment:
There can't be more than like 2^29 elements without proc macros as you only get 2^32 bytes for all source code combined and you need on average multiple characters for each identifier. In other words support for 2^35 elements is plenty.

Reply:
I think that's probably not true for e.g. AllocIds, right? At least with miri, you can presumably run a program that runs indefinitely and so allocates filling up memory here. You'd probably run out of RAM, etc., so I'm not actually particularly worried though :)

Reply:
Right

Reply:
In Miri (and CTFE), you can keep allocating and freeing memory and that way use up `AllocId` without using up more and mire RAM.

Reply:
... but if you do that, those allocations don't end up in `alloc_map`; only the final value of a const/static is put there.